ethereum/dbstack/db_stack.go

```diff
@@ -1,91 +1,168 @@
 package dbstack
 
 import (
+	"fmt"
 	"sync"
 
 	"github.com/0xProject/0x-mesh/ethereum/miniheader"
 	"github.com/0xProject/0x-mesh/meshdb"
 )
 
-// DBStack allows performing basic stack operations on a stack of meshdb.MiniHeaders stored
-// in the DB backing our meshdb package.
-type DBStack struct {
-	// TODO(albrow): Use Transactions when db supports them instead of a mutex
-	// here. There are cases where we need to make sure no modifications are made
-	// to the database in between a read/write or read/delete.
-	mut    sync.Mutex
-	meshDB *meshdb.MeshDB
-	limit  int
+type updateType int
+
+const (
+	pop updateType = iota
+	push
+)
+
+type update struct {
+	Type       updateType
+	MiniHeader *miniheader.MiniHeader
 }
 
-// New instantiates a new stack with the specified size limit. Once the size limit
-// is reached, adding additional blocks will evict the deepest block.
-func New(meshDB *meshdb.MeshDB, limit int) *DBStack {
-	return &DBStack{
-		meshDB: meshDB,
-		limit:  limit,
-	}
+// DBStack is an in-memory stack that can be sync'd with the DB
+type DBStack struct {
+	meshDB      *meshdb.MeshDB
+	limit       int
+	miniHeaders []*miniheader.MiniHeader
+	mu          sync.Mutex
+	updates     []*update
 }
 
-// Pop removes and returns the latest block header on the block stack. It
-// returns nil if the stack is empty.
-func (b *DBStack) Pop() (*miniheader.MiniHeader, error) {
-	b.mut.Lock()
-	defer b.mut.Unlock()
-	latestMiniHeader, err := b.meshDB.FindLatestMiniHeader()
+// New instantiates a new DBStack
+func New(meshDB *meshdb.MeshDB, retentionLimit int) (*DBStack, error) {
+	miniHeaders, err := meshDB.FindAllMiniHeadersSortedByNumber()
 	if err != nil {
 		return nil, err
 	}
-	if latestMiniHeader == nil {
+	d := &DBStack{
+		meshDB:      meshDB,
+		limit:       retentionLimit,
+		miniHeaders: miniHeaders,
+		updates:     []*update{},
+	}
+	return d, nil
+}
+
+// Peek returns the top of the stack
+func (d *DBStack) Peek() (*miniheader.MiniHeader, error) {
```

Comment:
Since `DBStack` now also holds miniheaders in memory it seems like it is a lot more similar to `SimpleStack`. If I understand correctly, the only real changes are in the `Checkpoint` and `Reset` methods. Maybe we could reduce some duplicated code by having `DBStack` own a `SimpleStack` as a (possibly embedded) struct field?

Reply:
Ended up deduping by making DBStack instantiate a SimpleStack within it.